File: cpp/recursion.cpp
// RECURSION
// Demonstrates recursive functions and backtracking in C++

#include <iostream>
using namespace std;

// Simple recursion - factorial
int factorial(int n) {
    if (n <= 1) {
        return 1;
    }
    return n * factorial(n - 1);
}

// Fibonacci sequence
int fibonacci(int n) {
    if (n <= 1) {
        return n;
    }
    return fibonacci(n - 1) + fibonacci(n - 2);
}

// Binary search with recursion
int binarySearch(int arr[], int left, int right, int target) {
    if (left > right) {
        return -1;  // Not found
    }
    
    int mid = (left + right) / 2;
    
    if (arr[mid] == target) {
        return mid;
    } else if (arr[mid] > target) {
        return binarySearch(arr, left, mid - 1, target);
    } else {
        return binarySearch(arr, mid + 1, right, target);
    }
}

// Backtracking - Generate all permutations
void permute(string str, int l, int r) {
    if (l == r) {
        cout << str << endl;
    } else {
        for (int i = l; i <= r; i++) {
            swap(str[l], str[i]);
            permute(str, l + 1, r);
            swap(str[l], str[i]);  // Backtrack
        }
    }
}

int main() {
    // Factorial
    cout << "Factorial of 5: " << factorial(5) << endl;
    cout << "Factorial of 10: " << factorial(10) << endl;
    
    // Fibonacci
    cout << "Fibonacci(6): " << fibonacci(6) << endl;
    cout << "Fibonacci(10): " << fibonacci(10) << endl;
    
    // Binary search
    int numbers[] = {1, 3, 5, 7, 9, 11, 13, 15};
    int size = 8;
    int found = binarySearch(numbers, 0, size - 1, 7);
    if (found != -1) {
        cout << "Found at index: " << found << endl;
    } else {
        cout << "Not found" << endl;
    }
    
    // Permutations
    cout << "Permutations of 'ABC':" << endl;
    permute("ABC", 0, 2);
    
    return 0;
}

File: rust/recursion.rs
// RECURSION
// Demonstrates recursive functions and backtracking

// Simple recursion - factorial
fn factorial(n: u32) -> u32 {
    if n <= 1 {
        1
    } else {
        n * factorial(n - 1)
    }
}

// Fibonacci sequence
fn fibonacci(n: u32) -> u32 {
    if n <= 1 {
        n
    } else {
        fibonacci(n - 1) + fibonacci(n - 2)
    }
}

// Binary search with recursion
fn binary_search(arr: &[i32], left: i32, right: i32, target: i32) -> i32 {
    if left > right {
        return -1;  // Not found
    }
    
    let mid = (left + right) / 2;
    let mid_usize = mid as usize;
    
    if arr[mid_usize] == target {
        mid
    } else if arr[mid_usize] > target {
        binary_search(arr, left, mid - 1, target)
    } else {
        binary_search(arr, mid + 1, right, target)
    }
}

// Backtracking - Generate all permutations
fn permute(s: &mut Vec<char>, l: usize, r: usize, result: &mut Vec<String>) {
    if l == r {
        result.push(s.iter().collect());
    } else {
        for i in l..=r {
            s.swap(l, i);
            permute(s, l + 1, r, result);
            s.swap(l, i);  // Backtrack
        }
    }
}

// Tree node structure
#[derive(Debug)]
struct TreeNode {
    val: i32,
    left: Option<Box<TreeNode>>,
    right: Option<Box<TreeNode>>,
}

impl TreeNode {
    fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

// Sum all values in a binary tree
fn tree_sum(root: &Option<Box<TreeNode>>) -> i32 {
    match root {
        None => 0,
        Some(node) => node.val + tree_sum(&node.left) + tree_sum(&node.right),
    }
}

// Count nodes in a tree
fn count_nodes(root: &Option<Box<TreeNode>>) -> i32 {
    match root {
        None => 0,
        Some(node) => 1 + count_nodes(&node.left) + count_nodes(&node.right),
    }
}

fn main() {
    // Factorial
    println!("Factorial of 5: {}", factorial(5));
    println!("Factorial of 10: {}", factorial(10));
    
    // Fibonacci
    println!("Fibonacci(6): {}", fibonacci(6));
    println!("Fibonacci(10): {}", fibonacci(10));
    
    // Binary search
    let numbers = vec![1, 3, 5, 7, 9, 11, 13, 15];
    let found = binary_search(&numbers, 0, numbers.len() as i32 - 1, 7);
    println!("Found 7 at index: {}", found);
    
    // Permutations
    let mut perms = Vec::new();
    let mut chars = vec!['A', 'B', 'C'];
    permute(&mut chars, 0, 2, &mut perms);
    println!("Permutations of 'ABC': {:?}", perms);
    
    // Tree operations
    let mut root = Box::new(TreeNode::new(1));
    root.left = Some(Box::new({
        let mut left = TreeNode::new(2);
        left.left = Some(Box::new(TreeNode::new(4)));
        Box::new(left)
    }));
    root.right = Some(Box::new({
        let mut right = TreeNode::new(3);
        right.right = Some(Box::new(TreeNode::new(5)));
        Box::new(right)
    }));
    
    let root_option = Some(root);
    println!("Tree sum: {}", tree_sum(&root_option));
    println!("Tree node count: {}", count_nodes(&root_option));
}
